### PR TITLE
Supporting more complicated case_of and other patterns

### PR DESCRIPTION
**Pull request by JaroslavTulach**
### Pull Request Description

Bigger alignment between Rust based and old parser.

### Checklist

Please include the following checklist in your PR:

- [ x ] All code conforms to the
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
- All code has been tested:
  - [ x ] Unit tests have been written where possible.

[ci no changelog needed]

### CODE REVIEW

File: engine/runtime/src/main/java/org/enso/compiler/TreeToIr.java [hunk 1]
```diff
@@ -646,6 +642,25 @@ yield switch (identifier) {
...
         yield fn;
       }
+      case Tree.NamedApp app -> {
+        var fn = translateExpression(app.getFunc(), cons(app, moreArgs), insideTypeSignature, false);
+        yield fn;
+      }
+      case Tree.Array arr -> {
+        List<IR.Expression> items = nil();
+        if (arr.getFirst() != null) {
+          var exp = translateExpression(arr.getFirst(), insideTypeSignature);
```
Comment (hubertp) [inline engine/runtime/src/main/java/org/enso/compiler/TreeToIr.java:652]: Shouldn't this be
```suggestion
          var exp = translateExpression(arr.getFirst().getBody(), insideTypeSignature);
```
? Or is that the reason for this getFirst vs getRest?

**Review by hubertp — APPROVED** [review]

File: engine/runtime/src/main/java/org/enso/compiler/TreeToIr.java (hunk 1 above)
Comment (JaroslavTulach) [reply engine/runtime/src/main/java/org/enso/compiler/TreeToIr.java:652]: I think the current state is OK. `arr.getFirst()` is just a `Tree`. Elements of `arr.getRest()` are `​OperatorDelimitedTree` and one needs to call `getBody` to get the `Tree` from them.

File: engine/runtime/src/main/java/org/enso/compiler/TreeToIr.java [hunk 2]
```diff
@@ -1068,26 +1095,104 @@ private static boolean isOperator(String txt, Either<MultipleOperatorError, Oper
...
+            Option.empty(),
+            isSuspended,
+            loc,
+            meta(), diag()
+          );
+      }
+      default -> throw new UnhandledEntity(core, "translateArgumentDefinition");
     };
-    boolean isSuspended = def.getSuspension() != null;
+    /*
```
Comment (kazcw) [inline engine/runtime/src/main/java/org/enso/compiler/TreeToIr.java:1195]: The old implementation was simpler. Is there a problem with it the tests don't catch?

Comment (JaroslavTulach) [reply engine/runtime/src/main/java/org/enso/compiler/TreeToIr.java:1195]: I merged with conflicts. I had to resolve them. Then I had to make the tests pass. This is the result.

**Pull request merged**